### PR TITLE
Initial optional configmap proposal

### PR DESCRIPTION
**Pull request by fraenkel**
Proposal for #170

### CODE REVIEW

**Comment by thockin** [comment]
LGTM

**Comment by thockin** [comment]
I'll leave this open for a few days to collect feedback

**Comment by liggitt** [comment]
Would like to keep the secret envFrom and volume spec consistent with whatever we do here

**Comment by fraenkel** [comment]
Yes, the intent would be to apply to others. Until the initial EnvFrom support goes in followed by add ons for Secrets, then Optional can apply to other bits. But right now, Secrets aren't there so it's cart before the horse.

File: contributors/design-proposals/optional-configmap.md [hunk 1]
```diff
@@ -0,0 +1,111 @@
...
+spec:
+  containers:
+  - name: etcd
+    image: openshift/etcd-20-centos7
+    ports:
+    - containerPort: 2379
+      protocol: TCP
+    - containerPort: 2380
+      protocol: TCP
+    envFrom:
```
Comment (liggitt) [inline contributors/design-proposals/optional-configmap.md]: need to define the behavior of an envvar that is both defined literally and from an optional config map which doesn't exist

```
...
spec:
  containers:
  - name: etcd
    image: openshift/etcd-20-centos7
    env:
    - name: foo
      value: bar
    - name: foo
      valueFrom:
        configMapKeyRef:
          name: configMapA
          key: some-key
          optional: true
```

**Comment by liggitt** [comment]
should also apply to env valueFrom configMapKeyRef/secretKeyRef

**Comment by fraenkel** [comment]
@liggitt Hmm, I hadn't thought of making configmaps optional everywhere. But from the consistency POV I can see the rationale.

File: contributors/design-proposals/optional-configmap.md [hunk 2]
```diff
@@ -0,0 +1,111 @@
...
+variables. These ConfigMaps can be marked as optional. The default behavior
+will cause the container to fail to start when the ConfigMap is not defined.
+If the ConfigMap is marked optional, when the ConfigMap is not defined, the
+container will skip this ConfigMap and continue.
+
+A pod can specify a set of Volumes to mount. A ConfigMap can represent the
+files to populate the volume. The ConfigMaps should be allowed to be optional.
+The default behavior will fail the volume setup if the ConfigMap is missing.
+If the ConfigMap is marked as optional, the volume will be set up with no
+files, and continue.
```
Comment (liggitt) [inline contributors/design-proposals/optional-configmap.md]: Currently, the files in the configmap volume are updated from the configmap while the pod is running. 

Define what happens if an optional configmap is missing when the pod is started on the node, and the configmap appears later while the pod is running... will the files be populated at that point, or not?

File: contributors/design-proposals/optional-configmap.md [hunk 3]
```diff
@@ -0,0 +1,111 @@
...
+container and files within a Volume to be optional.
+
+## Use Cases
+
+An application developer can define a set of environment variables for a
+container by referencing a ConfigMap with default values. The developer can
+also define an optional ConfigMap that is used to override the values. There
+may be a set of these optional ConfigMaps that provide environment specific
+overrides for a given test cluster or region cluster. If no overrides are
+present, the container will get the default values.
```
Comment (liggitt) [inline contributors/design-proposals/optional-configmap.md]: > the container will get the default values

Not sure what "default values" means here. Is this the "define literally, then define optional configmap overrides" case I asked about below?

**Comment by fraenkel** [comment]
@liggitt Hopefully this helps a bit more.

**Comment by liggitt** [comment]
thanks

File: contributors/design-proposals/optional-configmap.md (hunk 1 above)
Comment (thockin) [inline contributors/design-proposals/optional-configmap.md]: Agree.  This is a great corner case.  I think defining it as "unchanged from previous value" seems correct.

**Comment by thockin** [comment]
Let's get this feature implemented!

@MrHohn I know you want this...

**Comment by fruch** [comment]
This is now mentioned in the offical documention, but https://kubernetes.io/docs/api-reference/v1/definitions/ doesn't show them at all

**Comment by liggitt** [comment]
that site publishes docs for the current release, which is still 1.5.x

**Comment by fruch** [comment]
I find it very confusing that the official ConfigMap docs says something different than current realease. I would expect official documentation to mention which version introduced what feature clearly. I was supprized that optional didn't worked for, when trying to follow documentation

**Comment by liggitt** [comment]
What official docs are you referring to?

**Comment by fruch** [comment]
Those:
https://kubernetes.io/docs/user-guide/configmap/

**Comment by fruch** [comment]
I.e. the first hit when looking for ConfigMap...

**Comment by liggitt** [comment]
@fraenkel @jaredbhatti looks like the doc PR merged to master prematurely. Can we back those docs out and open the PR for the 1.6 doc branch instead?

**Comment by fraenkel** [comment]
Someone needs to create a release-1.6 branch.

**Comment by thockin** [comment]
@devin-donnelly

**Comment by fraenkel** [comment]
So how do we revert this?
I can rebase on the 1.6 branch now that it is there.

**Comment by fraenkel** [comment]
I will send a PR to revert.

**Comment by duffqiu** [comment]
#55368

**Pull request merged**